File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters/Session/Trackers/ChangeSetTracker.cs
// <copyright file="ChangeSetTracker.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>
//   Defines the AllorsChangeSetMemory type.
// </summary>

namespace Allors.Workspace.Adapters
{
    using System;
    using System.Collections.Generic;

    public sealed class ChangeSetTracker
    {
        public ChangeSetTracker(Workspace session) => this.Session = session;

        public Workspace Session { get; set; }

        public ISet<IStrategy> Created { get; set; }

        public ISet<IStrategy> Instantiated { get; set; }

        public ISet<DatabaseState> DatabaseOriginStates { get; set; }

        public void OnCreated(Strategy strategy)
        {
            (this.Created ??= new HashSet<IStrategy>()).Add(strategy);
            this.Session.OnChanged(EventArgs.Empty);
        }

        public void OnInstantiated(Strategy strategy)
        {
            (this.Instantiated ??= new HashSet<IStrategy>()).Add(strategy);
            this.Session.OnChanged(EventArgs.Empty);
        }

        public void OnDatabaseChanged(DatabaseState state)
        {
            (this.DatabaseOriginStates ??= new HashSet<DatabaseState>()).Add(state);
            this.Session.OnChanged(EventArgs.Empty);
        }
    }
}

File: typescript/libs/system/workspace/adapters/src/lib/session/trackers/change-set-tracker.ts
import { IObject } from '@allors/system/workspace/domain';
import { DatabaseState } from '../state/database-state';

export class ChangeSetTracker {
  created: Set<IObject>;
  databaseOriginStates: Set<DatabaseState>;

  public onCreated(object: IObject) {
    (this.created ??= new Set()).add(object);
  }

  public onDatabaseChanged(state: DatabaseState) {
    (this.databaseOriginStates ??= new Set()).add(state);
  }

  onDelete(object: IObject) {
    this.created?.delete(object);
  }
}
